File: day04/solution.py
"""Advent of Code 2025 - Day 4: Printing Department"""


def count_neighbors(grid: list[str], row: int, col: int) -> int:
    """Count how many @ symbols are adjacent to this cell (8 directions)."""
    rows = len(grid)
    cols = len(grid[0])
    count = 0

    # check all 8 directions: up, down, left, right, and diagonals
    for dr in [-1, 0, 1]:
        for dc in [-1, 0, 1]:
            if dr == 0 and dc == 0:
                continue  # skip the cell itself

            nr, nc = row + dr, col + dc

            # make sure we're still inside the grid
            if 0 <= nr < rows and 0 <= nc < cols:
                if grid[nr][nc] == '@':
                    count += 1

    return count


def find_accessible(grid: list[list[str]]) -> list[tuple[int, int]]:
    """Find all rolls that can be accessed (fewer than 4 adjacent rolls)."""
    accessible = []

    for row in range(len(grid)):
        for col in range(len(grid[0])):
            if grid[row][col] == '@':
                neighbors = count_neighbors(grid, row, col)
                if neighbors < 4:
                    accessible.append((row, col))

    return accessible


def solve(grid: list[str]) -> int:
    """Part 1: Count rolls accessible by forklift."""
    return len(find_accessible([list(row) for row in grid]))


def solve_part2(grid: list[str]) -> int:
    """Part 2: Keep removing accessible rolls until none left, count total removed."""
    # convert to mutable grid (list of lists instead of list of strings)
    grid = [list(row) for row in grid]
    total_removed = 0

    while True:
        accessible = find_accessible(grid)
        if not accessible:
            break  # no more rolls can be removed

        # remove all accessible rolls
        for row, col in accessible:
            grid[row][col] = '.'

        total_removed += len(accessible)

    return total_removed


# example from the puzzle
example_input = [
    "..@@.@@@@.",
    "@@@.@.@.@@",
    "@@@@@.@.@@",
    "@.@@@@..@.",
    "@@.@@@@.@@",
    ".@@@@@@@.@",
    ".@.@.@.@@@",
    "@.@@@.@@@@",
    ".@@@@@@@@.",
    "@.@.@@@.@.",
]


if __name__ == "__main__":
    with open("input.txt") as f:
        real_input = [line.strip() for line in f if line.strip()]

    # Part 1
    print("--- Part 1 ---")
    result = solve(example_input)
    print(f"Example: {result}")  # should be 13
    answer = solve(real_input)
    print(f"Answer:  {answer}")

    # Part 2
    print("\n--- Part 2 ---")
    result2 = solve_part2(example_input)
    print(f"Example: {result2}")  # should be 43
    answer2 = solve_part2(real_input)
    print(f"Answer:  {answer2}")

File: day04/solution.go
// Advent of Code 2025 - Day 4: Printing Department
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// countNeighbors counts how many @ symbols are adjacent (8 directions)
func countNeighbors(grid [][]byte, row, col int) int {
	rows := len(grid)
	cols := len(grid[0])
	count := 0

	// check all 8 directions - same nested loop trick as Python
	for dr := -1; dr <= 1; dr++ {
		for dc := -1; dc <= 1; dc++ {
			if dr == 0 && dc == 0 {
				continue // skip the cell itself
			}

			nr, nc := row+dr, col+dc

			// bounds check
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				if grid[nr][nc] == '@' {
					count++
				}
			}
		}
	}

	return count
}

// position is a simple struct to hold row,col - Go doesn't have tuples
type position struct {
	row, col int
}

// findAccessible returns all positions of rolls with < 4 neighbors
func findAccessible(grid [][]byte) []position {
	var accessible []position

	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid[0]); col++ {
			if grid[row][col] == '@' {
				neighbors := countNeighbors(grid, row, col)
				if neighbors < 4 {
					accessible = append(accessible, position{row, col})
				}
			}
		}
	}

	return accessible
}

// copyGrid creates a mutable copy of the grid
// [][]byte is a 2D slice - each row is a separate slice
func copyGrid(lines []string) [][]byte {
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		// []byte(line) converts string to byte slice (mutable)
		grid[i] = []byte(line)
	}
	return grid
}

// solve counts rolls accessible by forklift
func solve(lines []string) int {
	grid := copyGrid(lines)
	return len(findAccessible(grid))
}

// solvePart2 keeps removing accessible rolls until none left
func solvePart2(lines []string) int {
	grid := copyGrid(lines)
	totalRemoved := 0

	for {
		accessible := findAccessible(grid)
		if len(accessible) == 0 {
			break // no more to remove
		}

		// remove all accessible rolls
		for _, pos := range accessible {
			grid[pos.row][pos.col] = '.'
		}

		totalRemoved += len(accessible)
	}

	return totalRemoved
}

// loadInput reads non-empty lines from a file
func loadInput(filename string) []string {
	file, _ := os.Open(filename)
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			lines = append(lines, line)
		}
	}
	return lines
}

func main() {
	exampleInput := []string{
		"..@@.@@@@.",
		"@@@.@.@.@@",
		"@@@@@.@.@@",
		"@.@@@@..@.",
		"@@.@@@@.@@",
		".@@@@@@@.@",
		".@.@.@.@@@",
		"@.@@@.@@@@",
		".@@@@@@@@.",
		"@.@.@@@.@.",
	}

	// Part 1
	fmt.Println("--- Part 1 ---")
	result := solve(exampleInput)
	fmt.Printf("Example: %d\n", result) // should be 13

	realInput := loadInput("input.txt")
	answer := solve(realInput)
	fmt.Printf("Answer:  %d\n", answer)

	// Part 2
	fmt.Println("\n--- Part 2 ---")
	result2 := solvePart2(exampleInput)
	fmt.Printf("Example: %d\n", result2) // should be 43

	answer2 := solvePart2(realInput)
	fmt.Printf("Answer:  %d\n", answer2)
}
